Client.py

```diff
@@ -0,0 +1,227 @@
+        self._socket = s
+        self._recv_buff = ""
+        self._send_buff = ""
+        self._buff_size = buff_size
+        self.file = file_object
+        self.request = request
+
+    def get_socket(self):
+        """Get the socket of this client, private arguement"""
+        return self._socket
+
+    def _parse_header(self, line):
+        parsed = line.split(":", 1)
+        if len(parsed) == 2:
+            self.request.add_header(*parsed)
+            self.logger.debug("Added header, %s", line)
+
+    def recv(self):
+        """Recv data from the client socket and process it to Reqeust and
+           FileObject, or put it in the buff"""
+        try:
+            line = ""
+            # If there is no request, get request line
+            if self.request is None and constants.CRLF not in self._recv_buff:
+                self._recv_lines()
+                self.logger.debug("after recv %s", self._recv_buff)
+            # If we have the request line, but don't have request object, creat
+            # one
+            if self.request is None and constants.CRLF in self._recv_buff:
+                parsed_lines = self._recv_buff.split(constants.CRLF, 1)
+                req = parsed_lines[0].split(' ', 2)
+                self.logger.debug("Req, %s %s %s", req)
+                if req[2] != constants.HTTP_VERSION:
+                    raise RuntimeError('Not HTTP protocol')
+                if len(req) != 3:
+                    raise RuntimeError('Incomplete HTTP protocol')
+                method, uri, signature = req
+                if method not in SUPPORTED_METHODS:
+                    raise RuntimeError(
+                        "HTTP unspported method '%s'" % method)
+                if not uri or uri[0] != '/' or '\\' in uri:
+                    raise RuntimeError("Invalid URI")
+                file_name = os.path.normpath(
+                    '%s%s' % (constants.BASE, os.path.normpath(uri)))
+                self.file = FileObject.FileObject(file_name)
+                self.request = Request.Request(method, uri)
+                self._send_buff += (
+                    "%s 200 OK\r\n"
+                    "Content-Length: %s\r\n"
+                    "Content-Type: %s\r\n"
+                    "\r\n"
+                ) % (
+                    constants.HTTP_VERSION,
+                    self.file.get_file_size(),
+                    MIME_MAPPING.get(
+                        os.path.splitext(
+                            self.request.get_uri()
+                        )[1].lstrip('.'),
+                        'application/octet-stream',
+                    ),
+                )
+                self.logger.debug("Created file and request")
+                if len(parsed_lines) == 1:
+                    self._recv_buff = ""
+                else:
+                    self._recv_buff = parsed_lines[1]
+
+            # If we do have request line, get headers
+            if self.request is not None:
```

Comment:
this is 10000000 lines function!
please separate into smaller functions
please use state machine instead of these ugly edge conditions of magic variables.